core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,415 @@
+            tombstones.add(tombstone);
+            return tombstone;
+        }
+
+        /**
+         * Purge tombstone entries.  Returns the number of entries that were purged.
+         *
+         * Tombstones are purged if the number of tombstones in the list
+         * is greater than the input parameter of maximum allowed tombstones.
+         * Tombstones are purged until the list is equal to the maximum allowed.
+         */
+        public int purge(final int maxTombstones) {
+            int numToPurge = tombstones().size() - maxTombstones;
+            if (numToPurge <= 0) {
+                return 0;
+            }
+            tombstones = tombstones.subList(numToPurge, tombstones.size());
+            return numToPurge;
+        }
+
+        public IndexGraveyard build() {
+            return new IndexGraveyard(tombstones);
+        }
+    }
+
+    /**
+     * A class representing a diff of two IndexGraveyard objects.
+     */
+    final public static class IndexGraveyardDiff implements Diff<MetaData.Custom> {
+
+        private final List<Tombstone> added;
+        private final int removedCount;
+
+        IndexGraveyardDiff(final StreamInput in) throws IOException {
+            added = Collections.unmodifiableList(in.readList((streamInput) -> new Tombstone(streamInput)));
+            removedCount = in.readVInt();
+        }
+
+        IndexGraveyardDiff(final IndexGraveyard previous, final IndexGraveyard current) {
+            final List<Tombstone> previousTombstones = previous.tombstones;
+            final List<Tombstone> currentTombstones = current.tombstones;
+            final List<Tombstone> added;
+            int removed = 0;
+            if (previousTombstones.isEmpty()) {
+                // nothing will have been removed, and all entries in current are new
+                added = new ArrayList<>(currentTombstones);
+            } else if (currentTombstones.isEmpty()) {
+                // nothing will have been added, and all entries in previous are removed
+                added = Collections.emptyList();
+                removed = previousTombstones.size();
+            } else {
+                // look through the back, starting from the end, for added tombstones
+                final Tombstone lastAddedTombstone = previousTombstones.get(previousTombstones.size() - 1);
+                final int addedIndex = currentTombstones.lastIndexOf(lastAddedTombstone);
+                if (addedIndex < currentTombstones.size()) {
+                    added = currentTombstones.subList(addedIndex + 1, currentTombstones.size());
+                } else {
+                    added = Collections.emptyList();
+                }
+                // look through the front, starting at the beginning, for removed tombstones
+                final Tombstone firstTombstone = currentTombstones.get(0);
+                removed = Math.max(previousTombstones.indexOf(firstTombstone), 0);
```

Comment:
You are correct, I will fix this and add a specific test.  The tests are based on randomization for the number of removed and number of added, so I suspect it hasn't triggered the edge case yet.  I will add a specific test for that edge case.